sympy/matrices/common.py

```diff
@@ -2218,37 +2223,50 @@ def __neg__(self):
         return self._eval_scalar_mul(-1)
 
     @call_highest_priority('__rpow__')
-    def __pow__(self, num):
+    def __pow__(self, exp):
         if self.rows != self.cols:
             raise NonSquareMatrixError()
         a = self
-        jordan_pow = getattr(a, '_matrix_pow_by_jordan_blocks', None)
-        num = sympify(num)
-        if num.is_Number and num % 1 == 0:
+        jordan_pow = getattr (a, '_matrix_pow_by_jordan_blocks', None)
+        exp = sympify(exp)
+
+        if exp.is_zero:
+            return a._new(a.rows, a.cols, lambda i, j: int(i == j))
+        if exp == 1:
+            return a
+
+        if exp.is_Number and exp % 1 == 0:
             if a.rows == 1:
-                return a._new([[a[0]**num]])
-            if num == 0:
-                return self._new(self.rows, self.cols, lambda i, j: int(i == j))
-            if num < 0:
-                num = -num
+                return a._new([[a[0]**exp]])
+            if exp < 0:
+                exp = -exp
                 a = a.inv()
             # When certain conditions are met,
             # Jordan block algorithm is faster than
             # computation by recursion.
-            elif a.rows == 2 and num > 100000 and jordan_pow is not None:
+            elif a.rows == 2 and exp > 100000 and jordan_pow is not None:
                 try:
-                    return jordan_pow(num)
+                    return jordan_pow(exp)
                 except MatrixError:
                     pass
```

Comment:
The logic here can be reorganised. This is calling jordan_pow which we do below as well. Also this should be catching NonInvertibleMatrixError. Maybe the code for attempting jordan_pow should go before the code for using recursion so we don't duplicate this block.

Reply:
It catches MatrixError I believe to silently pass on to `_eval_pow_by_recursion()` if something other than noninvertible matrix happens, like some sort of internal error.

Reply:
I think it's catching it here for the same reason as NonInvertible is caught below. We did find some examples earlier in this PR that raised for non-invertible matrices under jordan_pow but were being correctly computed because of the recursion method. (I think that the explicit tests you've added for jordan_pow were because of that).

Reply:
I actually thought of removing that part since it is a very strange condition: `elif a.rows == 2 and exp > 100000` so it will probably almost never be executed. Much less is the need to optimize its placement. `_eval_pow_by_recursion()` raises on noninvertible anyway.

Reply:
Maybe it should be removed. It seems that that is there for speed reasons but probably only for the relatively uninteresting case of matrices with integer coefficients.

I'd be more inclined to worry about the impact on simplicity for symbolic matrices. Consider the difference between A and B in:
```
M = Matrix([[(1+x), (1-x)], [(1-x), (1+x)]])
A = M**100
B = (M**n).subs(n, 100)
```
Avoiding expression blowup in the symbolic case is a big task that needs working on for matrices.

Reply:
Yes it does get ugly, but now since this branch is stable and working well I would like to close it out before moving on to other things. I am thinking of adding a matrix log implementation.

Reply:
Take a look at #17221 (which was probably inspired by our discussions)